### ISSUE TITLE
Too much white space around graphs in 3d projection

### ISSUE THREAD

**Issue by pepe78** [matplotlib/matplotlib#19519]
When using 3d projection:

fig = plt.figure(figsize=(20,15))
ax = fig.gca(projection='3d')

and save the figure into png file, there is way too much white space from left right top and bottom, so when I want to add this graph into latex and use it in publication, I have to do post-processing of the image and remove unnecessary amount of blank space.

**Comment by ianhi** [comment]
Hi @pepe78 that sounds annoying indeed. However, I think that this can likely be resolved with the existing tools in matplotlib. Can you please post about this (with a picture of the output and a description of what you'd expect) on https://discourse.matplotlib.org/ which is the best place for getting help with matplotlib.

**Comment by pepe78** [comment]
Done (reported at https://discourse.matplotlib.org/). As far as "resolved with the existing tools in matplotlib", it would be nice to have by default less amount of blank space as it is done when not doing 3d projections (2d graphs). Thanks!

**Comment by jklymak** [comment]
I'm not sure why there is such a large margin around axes3d plots.  I have a sneaking suspicion that the margin was developed outside the subplot paradigm, which also have margins, and now we have two sets of margins.  Definitely when you add an axes directly via `Axes3d(fig)` (which we want to deprecate), the margins are smaller.

**Comment by pepe78** [comment]
This was great advice I got on discourse:

fig = plt.figure(figsize=(20,15), constrained_layout=True)
plt.savefig('1.png', bbox_inches='tight')

Left and right margins were minimal when saved like that (and top and bottom seem also smaller).

**Comment by jklymak** [comment]
I'll actually re-open, because to me, at least, there is still a mystery about why the margins are so much larger than other axes.

**Comment by pepe78** [comment]
That makes sense - as I said before, it would be nice to have less margins on sides by default. When doing 2d plots, this actually really works nicely by default.

**Comment by WeatherGod** [comment]
Yes, the original Axes3D didn't work within the subplot() framework. And subplots leave extra room for the static axes labels, meanwhile, Axes3D needed to account for extra space for its axes labels within the axes plotting area. Originally, this was fine because you didn't make Axes3D objects using subplots, so you didn't get allocated extra margins for axes labels. When it became possible to specify `projection='3d'` to a subplot call, this had the inadvertent effect of getting extra margin space that wasn't really needed.

I never did figure out a good solution for this that wouldn't act counter-intuitively. Perhaps Axes3D needs to be reworked such that it can interpret and express the margin parameters itself rather than letting it be handled by 2D-based assumptions?

**Comment by srkunze** [comment]
I stumbled over this issue as well while trying to plot a representation of the Riemann zeta function.

Zooming in just cuts off left and right of the plot. :-/

![image](https://user-images.githubusercontent.com/1389648/134722792-635474de-1299-4b06-b038-63c54031905e.png)

**Comment by astromancer** [comment]
Just a note that you can control the margin sizes to some extent using `gridspec_kws`.

```python
fig, ax = plt.subplots(subplot_kw={'projection': '3d'},
                       gridspec_kw=dict(top=1.07, bottom=0.02, left=0, right=1))
fig.savefig('axes3d.png', dpi=600, bbox_inches='tight')
```
![axes3d](https://user-images.githubusercontent.com/5623552/229286721-0d6465d8-0c3e-40ae-b3e8-7cb96e353675.png)

**Comment by cgadal** [comment]
Is there a current workaround for this ? I am having some problems with a figure similar to this situation:

```python

import matplotlib.pyplot as plt
import numpy as np

X = np.arange(-5, 5, 0.25)
Y = np.arange(-5, 5, 0.25)
X, Y = np.meshgrid(X, Y)
R = np.sqrt(X**2 + Y**2)
Z = np.sin(R)


fig, ax = plt.subplots(1, 1, figsize=(10, 5), subplot_kw=dict(projection="3d"))
plt.subplots_adjust(0, 0, 1, 1, 0, 0)


surf = ax.plot_surface(X, Y, Z, linewidth=0, antialiased=False)

ax.view_init(elev=20, azim=-45, roll=0)

ax.set_box_aspect([5, 1, 1])  # Aspect ratio
ax.set_aspect("auto")

plt.show()

```
Not that I do not want to use the bbox_inches='tight' when saving the plot, as the figure is already prepared at the right size to be inserted in a paper.

![Fig](https://github.com/user-attachments/assets/24cca424-b5fb-484c-9afa-b76ce3496adb)

**Comment by jklymak** [comment]
Yes, the workaround is stated above - add Axes3D manually.

**Comment by jklymak** [comment]
OK, my mistake.  I guess I don't know how you can fix this.  Maybe @scottshambaugh has some insight.  The overall issue is that you set the box_aspect.  I guess the algorithm is conservatice about guessing which direction to shrink based on the current view.

**Comment by AndrejHafner** [comment]
Is there any progress regarding topic? I'm having the same problem trying to fit the axis to the full width of the window - left and right sides just get cut off. I've tried many things, like the tight layout, adjusting the margins and other things, but none of them worked.

![Image](https://github.com/user-attachments/assets/e0fbc6b2-b768-49c1-b397-3d7c5e39e580)

**Comment by cgadal** [comment]
@AndrejHafner 

My current workaround is to play with the ax position `ax.set_position([-0.04, -0.32, 1.05, 1.1])` inside the figure.

If for some reason you are using this in a figure, this ax will overlap the other ones, so you might need to hide the background: `ax.set_facecolor("none")` 

if in this figure you are using `layout='constrained'`, you can do:

```
fig.canvas.draw()
fig.set_layout_engine("none")
axarr[1].set_position([-0.04, -0.32, 1.05, 1.1])
axarr[1].set_facecolor("none")
```

to fix axis positions from constrained layout, and then disable it to only change the ax position of you 3d figure.

**Comment by AndrejHafner** [comment]
> [@AndrejHafner](https://github.com/AndrejHafner)
> 
> My current workaround is to play with the ax position `ax.set_position([-0.04, -0.32, 1.05, 1.1])` inside the figure.
> 
> If for some reason you are using this in a figure, this ax will overlap the other ones, so you might need to hide the background: `ax.set_facecolor("none")`
> 
> if in this figure you are using `layout='constrained'`, you can do:
> 
> ```
> fig.canvas.draw()
> fig.set_layout_engine("none")
> axarr[1].set_position([-0.04, -0.32, 1.05, 1.1])
> axarr[1].set_facecolor("none")
> ```
> 
> to fix axis positions from constrained layout, and then disable it to only change the ax position of you 3d figure.

Is the `axarr` in your code the Axes3D? I'm not sure which axis you are actually setting the position for. I tried this approach, but I couldn't get the plot to become wider.

**Comment by cgadal** [comment]
@AndrejHafner 

 Yes, it is the `Axes3D`.

```python

import matplotlib.pyplot as plt
import numpy as np

X = np.arange(-5, 5, 0.25)
Y = np.arange(-5, 5, 0.25)
X, Y = np.meshgrid(X, Y)
R = np.sqrt(X**2 + Y**2)
Z = np.sin(R)

fig = plt.figure(figsize=(10, 5), layout="constrained")
ax = fig.add_subplot(111, projection="3d")

surf = ax.plot_surface(X, Y, Z, linewidth=0, antialiased=False)

ax.view_init(elev=20, azim=-45, roll=0)
ax.set_box_aspect([5, 1, 1])  # Aspect ratio


# fig.canvas.draw()
# fig.set_layout_engine("none")
# ax.set_position([-0.75, -0.75, 2.5, 2.5])
# ax.set_facecolor("none")

plt.savefig("Fig_2.png")
plt.show()

```

gives:

![Image](https://github.com/user-attachments/assets/ac20af2d-1cbf-4507-b3c0-2c68769e437c)

Uncommenting the commented lines gives:

![Image](https://github.com/user-attachments/assets/16f466b3-2985-4e29-93c4-e114e1be8bea)

The two first numbers in `ax.set_position` will control the position, and the two last ones the scaling factor.

**Comment by scottshambaugh** [comment]
There are two issues being talked about here:
1) The default 3D margins are really large (still an open question, and should be the focus of this issue).
2) 3D data is being clipped too aggressively. I have a separate issue for that here, and discussion on that should move over: https://github.com/matplotlib/matplotlib/issues/23450. I took a stab at fixing it today but my approach wasn't the right one. I believe the fix will be to expand the clip boxes for each of the 3D artists to the whole figure, since there's no real reason to truncate them.

**Comment by QuLogic** [comment]
@scottshambaugh I think for 1, you would need to do something like #29051, namely set the "actual" position of the Axes to correctly reflect the effective aspect ratio of the box that the 3D world covers.

**Comment by chiavez** [comment]
Not sure if this adds much, but I have the same problem: my figure has lots of white space on the bottom:

<img width=50% alt="Image" src="https://github.com/user-attachments/assets/67acf13b-b239-4cf2-952d-39fb25a7a210" />

`constrained_layout = True` and `bbox_inches='tight'` does not work.

I had to write code that automatically crops the generated figure. But this only works for bitmap, for vector formats it is quite a problem...

**Comment by QuLogic** [comment]
So the fix from #29051 is essentially: `self.set_aspect(self.axesLim.height / self.axesLim.width)` at the beginning of `draw`, and this automatically takes care of adjusting the current position to best fit. If we can essentially take the 3D world limits and project them to 2D, it should be possible to apply a similar correction here.

But note that there was some preparatory work there to ensure that the transforms remained correct when the aspect ratio was not square (as they had kind of assumed that to be the case.)
